### ISSUE TITLE
[css-ruby][css-inline][css-pseudo] Interaction of initial letter and ruby

### ISSUE THREAD

**Issue by frivoal** [w3c/csswg-drafts#654]
I believe it is [well defined](https://drafts.csswg.org/css-ruby-1/#ruby-layout) what happens if you do this:

``` css
span {
  font-size: 3em;
}
```

``` html
<p><ruby>
  <rb><span>今</span>日</rb>
  <rp>(</rp><rt>今日</rt><rp>)</rp>
<ruby></p>
```

On the other hand, I am not sure it is defined what happens if you do that:

``` css
span {
  initial-letter: 3 2;
}
```

``` html
<p><ruby>
  <rb><span>今</span>日</rb>
  <rp>(</rp><rt>今日</rt><rp>)</rp>
<ruby></p>
```

On the one hand, these parts of https://drafts.csswg.org/css-ruby-1/#ruby-layout should still work:

> Each ruby base container is sized and positioned to contain exactly all of its ruby bases’ margin boxes

and

> Ruby annotation containers are stacked outward over or under their corresponding ruby base container, without any intervening space. 

On the other hand, I am not sure this one does:

> When a ruby structure is laid out, its base level is laid out on the line, aligned according to its vertical-align property exactly as if its ruby bases were a regular sequence of inline boxes.

Does it mean initial-letter applies normally? Presumably, if it does, then we can then apply the previously quoted rules, and the ruby base container, by including all of its ruby bases’ margin boxes includes the full size of the initial letter, and from there, positioning the ruby annotation containers works as usual.

Or does it instead negate the effect of initial-letter?

I don't think this is a case authors would typically do intentionally (although who knows, there is no bound to author creativity). But regardless, this needs to be defined. Also, this, which is the same, might occur by accident:

``` css
p::first-letter {
  initial-letter: 3 2;
}
```

``` html
<p><ruby>
  <rb>今日</rb>
  <rp>(</rp><rt>今日</rt><rp>)</rp>
<ruby></p>
```

As far as I can tell from [css-pseudo](https://drafts.csswg.org/css-pseudo/#selectordef-first-letter), ::first-letter would select 今, and we're back in the same situation.
(with that said, implementations are inconsistent about that: FF and IE/Edge do select it, but chrome and safari don't).

**Comment by frivoal** [comment]
Oops, I made a typo in the first comment
`  <rb><span>今日</rb>`
is supposed to be
`  <rb><span>今</span>日</rb>`

I'll edit in place, but for those getting notifications by email, please note the correction..

**Comment by upsuper** [comment]
IIRC WebKit/Blink treat ruby like inline block, so it is internally consistent for them. I'm not quite confident on what Gecko would do in this case, though. Do you think this is a common pattern to have initial letter with ruby?

**Comment by frivoal** [comment]
> Do you think this is a common pattern to have initial letter with ruby?

No, I don't. I've made some brief research on this topic, and hardly anything turned up. To me, it means we can pick whatever is simpler to interoperably implement, without worrying too much about the use case, as it is a fringe one.

I still think we should get rid of ambiguities in the spec. Ideally, I'd like to define a single normative behavior. If we cannot easily agree on what the behavior should be, I could live with explicitly stating this is undefined, but I would prefer not leaving the spec in a state that leaves you scratching your head wondering if it is defining it or or.

**Comment by upsuper** [comment]
Then I'd suggest we make it undefined and encourage people to feedback on the behavior here with real use cases. Currently browsers have different impl approach for ruby, so it may be challenging to define an interoperable behavior which is easy for everyone to follow at present.

**Comment by frivoal** [comment]
Sounds good. I though there ought to be uses cases for that, and I looked for them in printed material, and failed. So making this explicitely undefined sounds OK.

So, what would making this undefined entail?

Do we just add a sentence to https://drafts.csswg.org/css-ruby-1/#ruby-layout saying "How this interacts with the initial-letter property is undefined in this level", or do we try to pinpoint the undefinedness to something more more specific?

The way I see it, the problem is in :

[issue 1 in the ruby spec](https://drafts.csswg.org/css-ruby-1/#issue-ca88244c):

> Are internal ruby boxes inline-level?

and the way it combines with the [apply line from the initial-letter property](https://drafts.csswg.org/css-inline/#sizing-drop-initials):

> Applies to: 	::first-letter pseudo-elements and inline-level first child of a block container

as well as this sentence:

> When a ruby structure is laid out, its base level is laid out on the line, aligned according to its vertical-align property exactly as if its ruby bases were a regular sequence of inline boxes.


If “a regular sequence of inline boxes” means that ruby bases are inline-level boxes, then initial-letter should applies. Then we have to determine:
* Is the initial-letter box is a child box of the inline-level box it comes from? (is that the thing we leave undefined?)
  * If yes, how does it influence the margin box of that inline-level box? As far as I can tell from the initial-letter spec, the margin box of the initial letter itself is well defined, but that's not the same question.
  * If the initial-letter box is not a child of the inline-level box it comes from, then presumably it is a sibling, and presumably that box isn't inline level. Is that right? what happens to this box when it is a child of a ruby ruby-base-container box? [step 2 of Anonymous Ruby Box Generation](https://drafts.csswg.org/css-ruby-1/#anon-gen-inlinize) says it becomes display:inline or display:inline-block. Does that cancel out the effects of the initial-letter property? (is that the thing we leave undefined?)
    * If it does not, how does [step 3](https://drafts.csswg.org/css-ruby-1/#anon-gen-bare-inlines) work? What's the implication of wrapping a (not inline-level) initial letter box and the rest of the inline-level box it came from into an anonymous inline-level box on the exclusion effect of the initial letter? (is that a thing we leave undefined as well?)

**Comment by frivoal** [comment]
Entangling that mess tends to make me think we should actually define that initial-letter does not apply to children of ruby base container boxes and of ruby container boxes (and of ruby annotation container boxes as well, but I think the apply line from the initial-letter property already excludes these).

While it looked sort of reasonable to me at first that initial-letter on the first letter of a ruby base could be made to work just as well as setting a larger font size on that letter, that only seems to be true as long as `ruby-position` is whichever of  `over` or `under` that matches `block-start`. On the `block-end` side it doesn't seem to make a whole lot of sense.

**Comment by upsuper** [comment]
I think we should actually delay applying `initial-letter` to inline child to the next level. As far as I know, WebKit only supports `initial-letter` on `::first-letter`, and Gecko plans to do that in the first impl as well.

And the other question is how `::first-letter` interacts with ruby, then...

**Comment by frivoal** [comment]
Agenda+, to decide if we should leave the interaction of initial-letter and ruby undefined, or if we should define  initial-letter not to apply in ruby.

**Comment by frivoal** [comment]
What is mysterious about the interaction of `::first-letter` with ruby? Other than the fact that some properties, like `initial-letter` might not work on children/descendants of ruby (which seems to be unrelated to whether that descendant is a good old element or a `::first-letter`), I don't see what the problem is.

Don't misunderstand me, I am not saying I am sure it is OK. `::first-letter` is weird enough that there may very well be something tricky. I'm just not seeing what.

**Comment by upsuper** [comment]
Oh, hmmm, I guess you're right that there doesn't seem to be problem for interaction of `::first-letter` with ruby itself. Some propertiess involving may cause unexpected behavior, but that is nothing about the pseudo-element itself.

What was in my mind is the float-based usage of `::first-letter` for simulating the effect of `initial-letter`. I think the behavior is well-defined for that case, although the result is probably not what people may expect.

**Comment by frivoal** [comment]
Resolved that we will make it explicitely undefined whether initial letter applies to children of ruby base container boxes and of ruby container boxes. Nobody things this is something we should/can solve, but for now we'd rather leave the door open in case someone wants to look into it.

**Comment by dauwhe** [comment]
Closed via 50b0f6f2419a0d22ca1ab7cb102ef9ba9bd96afd.
